qa/L0_python_api/test_kserve.py

```diff
@@ -0,0 +1,234 @@
+        server = TestingUtils.setup_server()
+        # Current flow is KServeHttp.Options or KServeGrpc.Options have to be
+        # provided to ensure type and range validation occurs.
+        with pytest.raises(
+            tritonserver.InvalidArgumentError,
+            match="Incorrect type for options. options argument must be of type",
+        ):
+            frontend.Server(server, {"port": 8001})
+
+        TestingUtils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend", [HTTP_ARGS[0], GRPC_ARGS[0]])
+    def test_server_service_order(self, frontend):
+        server = TestingUtils.setup_server()
+        service = TestingUtils.setup_service(server, frontend)
+
+        TestingUtils.teardown_server(server)
+        TestingUtils.teardown_service(service)
+
+    @pytest.mark.parametrize("frontend, client_type", [HTTP_ARGS[:2], GRPC_ARGS[:2]])
+    def test_service_custom_port(self, frontend, client_type):
+        server = TestingUtils.setup_server()
+        options = frontend.Options(port=8005)
+        service = TestingUtils.setup_service(server, frontend, options)
+        client = TestingUtils.setup_client(client_type, url="localhost:8005")
+
+        # Confirms that service starts at port 8005
+        client.is_server_ready()
+
+        TestingUtils.teardown_client(client)
+        TestingUtils.teardown_service(service)
+        TestingUtils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend, client_type, url", [HTTP_ARGS, GRPC_ARGS])
+    def test_inference(self, frontend, client_type, url):
+        server = TestingUtils.setup_server()
+        service = TestingUtils.setup_service(server, frontend)
+
+        # TODO: use common/test_infer
+        assert TestingUtils.send_and_test_inference_identity(client_type, url=url)
+
+        TestingUtils.teardown_service(service)
+        TestingUtils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend, client_type, url", [HTTP_ARGS])
+    def test_http_req_during_shutdown(self, frontend, client_type, url):
+        server = TestingUtils.setup_server()
+        http_service = TestingUtils.setup_service(server, frontend)
+        http_client = client_type.InferenceServerClient(url=url)
+        model_name = "delayed_identity"
+        delay = 2  # seconds
+        input_data0 = np.array([[delay]], dtype=np.float32)
+
+        input0 = client_type.InferInput("INPUT0", input_data0.shape, "FP32")
+        input0.set_data_from_numpy(input_data0)
+
+        inputs = [input0]
+        outputs = [client_type.InferRequestedOutput("OUTPUT0")]
+
+        async_request = http_client.async_infer(
+            model_name=model_name, inputs=inputs, outputs=outputs
+        )
+        # http_service.stop() does not use graceful shutdown
+        TestingUtils.teardown_service(http_service)
+
+        # So, inference request will fail as http endpoints have been stopped.
+        with pytest.raises(InferenceServerException):
```

Comment:
What is the expected exeption message?

Reply:
Added checks that verify the exception message for each known exception being thrown in `TestKServe`